lib/ex_aws/s3/download.ex

```diff
@@ -0,0 +1,48 @@
+alias Experimental.GenStage
+
+defmodule ExAws.S3.Download do
+  @moduledoc false
+
+  defstruct [
+    :bucket,
+    :path,
+    :dest,
+    opts: [],
+  ]
+end
+
+defimpl ExAws.Operation, for: ExAws.S3.Download do
+
+  alias ExAws.S3.Download
+
+  def perform(op, config) do
+    {:ok, source} =
+      op.bucket
+      |> get_file_size(op.path, config)
+      |> Download.Source.start_link
+
+    {:ok, sink} = Download.Sink.start_link(op.dest)
+
+    for _ <- 1..Keyword.get(op.opts, :max_concurrency, 8) do
+      {:ok, worker} = Download.Worker.start_link(%{bucket: op.bucket, path: op.path, config: config})
+
+      GenStage.sync_subscribe(sink, to: worker, min_demand: 0, max_demand: 1)
+      GenStage.sync_subscribe(worker, to: source, min_demand: 0, max_demand: 1)
+    end
+
+    # how do I block till this is done?
```

Comment:
@josevalim This worked out super well, and using GenStage really simplified stuff. The Source just emits byte delimiters that need to be downloaded, workers use those to actually pull down the data, and then the chunks are folded back in to process which handles writing them out to file.

What I'm not clear on is how I might block until it's done. I looked into just returning `{:stop, ...etc}` from the producer when it reached the end of the byte chunks but that seemed to make all the subscribers and definite sink quit prior to actually downloading / writing all the chunks.